### PR TITLE
Backport of most bug fixes from PR #1949

### PR DESCRIPTION
**Pull request by jtramm**
This PR incorporates a number of bug fixes from PR #1949 from OpenMC develop. One of the fixes was necessary to make RVS work consistently across all architectures (AMD cards in particular would not work with RVS enabled without this fix). 

I attempted to cherry pick the commits by @ojschumann first, but ran into issues applying them given some of the code delta between our offloading version and the current develop branch, so replicating the content ended up being easier.

### CODE REVIEW

File: src/photon.cpp [hunk 1]
```diff
@@ -248,7 +248,7 @@ PhotonInteraction::PhotonInteraction(hid_t group)
   // Create Compton profile CDF
   n_profile_ = data::compton_profile_pz_size;
   profile_cdf_ = xt::empty<double>({n_shell, n_profile_});
-  for (int i = 0; i < n_shell; ++i) {
+  for (int i = 0; i < profile_pdf_.shape(0); ++i) {
```
Comment (paulromano) [inline src/photon.cpp]: Note that I also added a bug fix in this block of code in openmc-dev/openmc#2297. Probably want to backport that one as well while you're at it.

File: tests/regression_tests/triso/results_true.dat [hunk 2]
```diff
@@ -1,2 +1,2 @@
 k-combined:
-1.701412E+00 3.180881E-02
+1.701412E+00 3.180881E-0
```
Comment (paulromano) [inline tests/regression_tests/triso/results_true.dat]: Was this change intentional?

**Review by paulromano — CHANGES_REQUESTED** [review]
Thanks for backporting these! I guess this is the price we pay for living in the past...

File: tests/regression_tests/triso/results_true.dat (hunk 2 above)
Comment (jtramm) [reply tests/regression_tests/triso/results_true.dat]: Definitely a mistake -- I had updated the numbers following the changes made in PR, but then decided not to as the new numbers would have been generated using the new PRNG, so would not correspond to the current PRNG anyhow. Meant to roll back my change but must have clipped the 2 at the end. Great catch!!!

File: src/photon.cpp (hunk 1 above)
Comment (jtramm) [reply src/photon.cpp]: Thanks for letting me know about that one! I've added in the changes from that PR as well.

**Review by paulromano — APPROVED** [review]
